test/unit-test/core_mqtt_utest.c

```diff
@@ -308,6 +356,241 @@ static int32_t transportWritevSuccess( NetworkContext_t * pNetworkContext,
+    uint8_t serializedLength[ 2 ];
+    const uint8_t * buffer;
+    size_t length;
+
+    serializedLength[ 0 ] = ( uint8_t ) ( topicFilterLength >> 8 );
+    serializedLength[ 1 ] = ( uint8_t ) ( topicFilterLength & 0x00ffU );
+
+    buffer = pIoVectorIterator[ 0 ].iov_base;
+    length = pIoVectorIterator[ 0 ].iov_len;
+
+    TEST_ASSERT_EQUAL( 2U, length );
+    TEST_ASSERT_EQUAL_UINT8_ARRAY( serializedLength, buffer, 2U );
+    ( *pBytesToWrite ) += length;
+
+    buffer = pIoVectorIterator[ 1 ].iov_base;
+    length = pIoVectorIterator[ 1 ].iov_len;
+
+    /* Encoded topic string. */
+    TEST_ASSERT_EQUAL( topicFilterLength, length );
+    TEST_ASSERT_EQUAL_UINT8_ARRAY( pTopicFilter, buffer, topicFilterLength );
+    ( *pBytesToWrite ) += length;
+
+    buffer = pIoVectorIterator[ 2 ].iov_base;
+    length = pIoVectorIterator[ 2 ].iov_len;
+
+    /* Encoded QoS. */
+    TEST_ASSERT_EQUAL( 1U, length );
+    TEST_ASSERT_EQUAL_UINT8( topicQos, *buffer );
+    ( *pBytesToWrite ) += length;
+}
+
+static void verifyEncodedTopicStringUnsubscribe( TransportOutVector_t * pIoVectorIterator,
+                                                 char * pTopicFilter,
+                                                 size_t topicFilterLength,
+                                                 int32_t * pBytesToWrite )
+{
+    /* Encoded topic length. */
+    uint8_t serializedLength[ 2 ];
+    const uint8_t * buffer;
+    size_t length;
+
+    serializedLength[ 0 ] = ( uint8_t ) ( topicFilterLength >> 8 );
+    serializedLength[ 1 ] = ( uint8_t ) ( topicFilterLength & 0x00ffU );
+
+    buffer = pIoVectorIterator[ 0 ].iov_base;
+    length = pIoVectorIterator[ 0 ].iov_len;
+
+    TEST_ASSERT_EQUAL( 2U, length );
+    TEST_ASSERT_EQUAL_UINT8_ARRAY( serializedLength, buffer, 2U );
+    ( *pBytesToWrite ) += length;
+
+    buffer = pIoVectorIterator[ 1 ].iov_base;
+    length = pIoVectorIterator[ 1 ].iov_len;
+
+    /* Encoded topic string. */
+    TEST_ASSERT_EQUAL( topicFilterLength, length );
+    TEST_ASSERT_EQUAL_UINT8_ARRAY( pTopicFilter, buffer, topicFilterLength );
+    ( *pBytesToWrite ) += length;
+}
+
+/**
+ * @brief Mocked transport writev to verify correct subscription
+ * packet.
+ *
+ * @param[in] tcpSocket TCP socket.
+ * @param[in] pMessage vectors to send
+ * @param[in] bytesToWrite number of vectors
+ *
+ * @return Number of bytes sent; negative value on error;
+ * 0 for timeout or 0 bytes sent.
+ */
+static int32_t transportWritevSubscribeSuccess( NetworkContext_t * pNetworkContext,
+                                                TransportOutVector_t * pIoVectorIterator,
+                                                size_t vectorsToBeSent )
+{
+    TEST_ASSERT_EQUAL( MQTT_SAMPLE_NETWORK_CONTEXT, pNetworkContext );
+    int32_t bytesToWrite = 0;
+    static int writeCount = 0;
+    const uint8_t * buffer;
+    size_t length;
+
+    /* The header. */
+    if( writeCount == 0 )
+    {
+        TEST_ASSERT_EQUAL( 4, vectorsToBeSent );
```

Comment:
Because the whole packet is not sent in one go. It is sent in chunks in successive calls to the `writev` function.
Thus, we need to verify that the packet is correct in chunks.